R/EventStudy.R

```diff
@@ -170,56 +174,90 @@ EventStudy <- function(estimator, data, outcomevar, policyvar, idvar, timevar, c
+        
+        return(any(holes_per_id$holes))
+    }
+
+    if (detect_holes(data, idvar, timevar)) {
+        warning(paste0("Note: gaps of more than one in the time variable '", timevar, "' were detected. ",
+                       "Treating these as gaps in the panel."))
+        timevar_holes <- TRUE
+    } else {
+        timevar_holes <- FALSE
+    }
+
+    if (post == 0 & overidpost == 0 & pre == 0 & overidpre == 0) {
+        static <- TRUE
+    } else {
+        static <- FALSE
     }
 
     num_evenstudy_coeffs <- overidpre + pre + post + overidpost
     num_periods          <- max(data[[timevar]], na.rm = T) - min(data[[timevar]], na.rm = T)
     if  (num_evenstudy_coeffs > num_periods - 1) {stop("overidpre + pre + post + overidpost cannot exceed the data window.")}
 
-    if  (sum(grepl(paste0(policyvar, "_fd"), colnames(data))) > 0) {warning(paste0("Variables starting with ", policyvar,
-                                                                                   "_fd should be reserved for eventstudyr."))}
-    if  (sum(grepl(paste0(policyvar, "_lead"), colnames(data))) > 0) {warning(paste0("Variables starting with ", policyvar,
-                                                                                     "_lead should be reserved for eventstudyr."))}
-    if  (sum(grepl(paste0(policyvar, "_lag"), colnames(data))) > 0) {warning(paste0("Variables starting with ", policyvar,
-                                                                                    "_lag should be reserved for eventstudyr."))}
+    for (tag in c("_fd", "_lead", "_lag")) {
+        if  (sum(grepl(paste0(policyvar, tag), colnames(data))) > 0) {
+            warning(paste0("Variables starting with ", policyvar, tag,
+                           " should be reserved for usage by eventstudyr."))
+        }
+    }
 
-    num_fd_lag_periods   <- post + overidpost - 1
-    num_fd_lead_periods  <- pre + overidpre
+    # Compute shifts in policy variable
+    num_fd_lags  <- post + overidpost - 1
+    num_fd_leads <- pre  + overidpre
 
-    furthest_lag_period  <- num_fd_lag_periods + 1
+    furthest_lag_period <- num_fd_lags + 1
 
-    if (post + overidpost + pre + overidpre > 0) {
-        data <- GetFirstDifferences(df = data, groupvar = idvar, timevar, diffvar = policyvar)
-    }
+    if (static) {
+        message("post, overidpost, pre, and overidpre are set to 0. A static model will be estimated.")
+    } else {
+        data <- ComputeFirstDifferences(data, idvar, timevar, policyvar, timevar_holes)
 
-    if (post + overidpost - 1 >= 1) {
-        data <- PrepareLags(data, groupvar = idvar, timevar,
-                            lagvar = paste0(policyvar, "_fd"), lags = 1:num_fd_lag_periods)
-    }
+        if ((post + overidpost - 1 >= 1) & (pre + overidpre >= 1)) {
+            shift_values = c(-num_fd_leads:-1, 1:num_fd_lags)
+        }
+        else if (pre + overidpre < 1) {
+            shift_values = 1:num_fd_lags
```

Comment:
@santiagohermo can you confirm this is right? I have a note to myself that lag = post = M and lead = pre = G. Of course my note could be wrong :) but just want to make sure that our "pre" variables go with lags and our "post" variables go with leads?

Reply:
It can be confusing, I agree @nateschor! While the names of objects and the order of code changed a bit, this should not have changed from before.

Here is the complete if
https://github.com/JMSLab/eventstudyr/blob/e43913a91d0668db61cba6bc8ecd340009466f10/R/EventStudy.R#L229-L237

If the user requests both post and pre, we use both leads and lags. If there only request for post values (so `pre + overidpre < 1`), we use only lags, and if only pre values (so `post + overidpost - 1 < 1`) we use only leads. This is similar to [before](https://github.com/JMSLab/eventstudyr/blob/fb7edc815c626c138b2d25f4cd8271411d15b91e/R/EventStudy.R#L195-L203), where with `post + overidpost - 1 >= 1` we computed lags and `pre + overidpre >= 1` we computed leads.

I think this makes sense. When we have `post + overidpost - 1 >= 1` we want to learn about dynamic effects, and to do that we need the values of the policy in the past (lags) to see how they affect today.

Would you agree @nateschor?

Reply:
Thanks @santiagohermo! This helps, and I agree